debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/LogMinerQueryResultProcessor.java

```diff
@@ -0,0 +1,245 @@
+                }
+                continue;
+            }
+
+            // Rollback
+            if (operationCode == RowMapper.ROLLBACK) {
+                if (transactionalBuffer.rollback(txId, logMessage)) {
+                    LOGGER.trace("ROLLBACK, {}", logMessage);
+                    rollbackCounter++;
+                }
+                continue;
+            }
+
+            // DDL
+            if (operationCode == RowMapper.DDL) {
+                // todo: DDL operations are not yet supported during streaming while using Log Miner.
+                LOGGER.info("DDL: {}, REDO_SQL: {}", logMessage, redo_sql);
+                continue;
+            }
+
+            // MISSING_SCN
+            if (operationCode == RowMapper.MISSING_SCN) {
+                LogMinerHelper.logWarn(transactionalBufferMetrics, "Missing SCN,  {}", logMessage);
+                continue;
+            }
+
+            // DML
+            if (operationCode == RowMapper.INSERT || operationCode == RowMapper.DELETE || operationCode == RowMapper.UPDATE) {
+                LOGGER.trace("DML,  {}, sql {}", logMessage, redo_sql);
+                dmlCounter++;
+                metrics.incrementCapturedDmlCount();
+                iterationStart = Instant.now();
+                LogMinerDmlEntry dmlEntry = dmlParser.parse(redo_sql, schema.getTables(), txId);
+                cumulativeParseTime = cumulativeParseTime.plus(Duration.between(iterationStart, Instant.now()));
+                iterationStart = Instant.now();
+
+                if (dmlEntry == null || redo_sql == null) {
+                    LOGGER.trace("Following statement was not parsed: {}, details: {}", redo_sql, logMessage);
+                    continue;
+                }
+
+                // this will happen for instance on a blacklisted column change, we will omit this update
+                if (dmlEntry.getCommandType().equals(Envelope.Operation.UPDATE)
+                        && dmlEntry.getOldValues().size() == dmlEntry.getNewValues().size()
+                        && dmlEntry.getNewValues().containsAll(dmlEntry.getOldValues())) {
+                    LOGGER.trace("Following DML was skipped, " +
+                            "most likely because of ignored blacklisted column change: {}, details: {}", redo_sql, logMessage);
+                    continue;
+                }
+
+                dmlEntry.setObjectOwner(segOwner);
+                dmlEntry.setSourceTime(changeTime);
+                dmlEntry.setTransactionId(txId);
+                dmlEntry.setObjectName(tableName);
+                dmlEntry.setScn(scn);
+
+                try {
+                    TableId tableId = RowMapper.getTableId(catalogName, resultSet);
+                    transactionalBuffer.registerCommitCallback(txId, scn, changeTime.toInstant(), redo_sql, (timestamp, smallestScn, commitScn, counter) -> {
```

Comment:
Why is there a commit callback for every record? With 100.000 records in a transaction this results in many callbacks... That probably all are very much the same. Why not just store a meta record at the start of a transaction (first occurrence of the idx) and the add only the sql to the list for that transaction?

Reply:
Also how do you handle very large transactions? I don't see optimizations for the memory usage.

Reply:
> Why is there a commit callback for every record? With 100.000 records in a transaction this results in many callbacks... That probably all are very much the same. Why not just store a meta record at the start of a transaction (first occurrence of the txId) and the add only the sql to the list for that transaction?

I most definitely agree that this could be reworked and far more memory friendly.  I've logged https://issues.redhat.com/browse/DBZ-2459 as a follow-up for this.

Reply:
> Also how do you handle very large transactions? I don't see optimizations for the memory usage.

I assume you're referring to the situation where the maps/sets grow but any excess memory that may be consumed by large or long running transactions is never recovered once they've been committed?

Reply:
In a high traffic environment (4000 mutations/sec) with large transactions of up to 100.000 in size, keeping a list with mutations and throwing the list away after use will introduce a lot of GC. So I think we have to be very much aware of optimizing the transaction storage and probably even find a way that GC does not get involved. So maybe some kind of custom memory management... I don't know how to do that yet. I'm thinking in the direction of using a dictionary with the most common words and referencing them. Column names for example probably are very stable and duplicated in every mutation.

Reply:
I got rid of most of the GC issues by using a charArray and reusing this for string storage. I think that something like this also has to be done in this connector. Otherwise it will never be able to handle big load with large transactions.

Reply:
@MilovdZee, could you add this as a comment to the Jira issue created by @Naros and linked above?

Reply:
done